### PR TITLE
Implement git_submodule_set_branch.

### PR DESCRIPTION
**Pull request by pks-t**

### CODE REVIEW

File: src/submodule.c [hunk 1]
```diff
@@ -715,6 +720,22 @@ const char *git_submodule_branch(git_submodule *submodule)
...
+	assert(submodule);
+
+	git__free(submodule->branch);
+
+	if (branch != NULL) {
+		submodule->branch = git__strdup(branch);
+		GITERR_CHECK_ALLOC(submodule->branch);
+	} else {
+		submodule->branch = NULL;
+	}
```
Comment (carlosmn) [inline src/submodule.c]: If the strdup fails, we will leave `submodule->branch` pointing to unallocated memory. We should set it to `NULL` always after freeing.

File: src/submodule.c [hunk 2]
```diff
@@ -628,8 +628,13 @@ int git_submodule_save(git_submodule *submodule)
 		(error = git_config_file_set_string(mods, key.ptr, submodule->url)) < 0)
 		goto cleanup;
 
-	if ((error = submodule_config_key_trunc_puts(&key, "branch")) < 0 ||
-		(error = git_config_file_set_string(mods, key.ptr, submodule->branch)) < 0)
+	if ((error = submodule_config_key_trunc_puts(&key, "branch")) < 0)
+		goto cleanup;
+	if (submodule->branch == NULL)
+		error = (error = git_config_file_delete(mods, key.ptr)) == GIT_ENOTFOUND ? 0 : error;
+	else
```
Comment (carlosmn) [inline src/submodule.c]: This is too much gymnastics with the error code. I'd rather see the style we have in the rest of the libarary

```C
error = git_config_file_delete(...);
if (error == GIT_ENOTFOUND)
    error = 0;
```

**Comment by pks-t** [comment]
Any comments/issues regarding this PR or could it be merged?

**Comment by pks-t** [comment]
Dang, wrong button.

**Comment by pks-t** [comment]
CI is having a good run, triggering rebuild.

File: src/submodule.c [hunk 3]
```diff
@@ -628,8 +628,15 @@ int git_submodule_save(git_submodule *submodule)
...
 
-	if ((error = submodule_config_key_trunc_puts(&key, "branch")) < 0 ||
-		(error = git_config_file_set_string(mods, key.ptr, submodule->branch)) < 0)
+	if ((error = submodule_config_key_trunc_puts(&key, "branch")) < 0)
+		goto cleanup;
+	if (submodule->branch == NULL)
+		error = git_config_file_delete(mods, key.ptr);
+	else
+		error = git_config_file_set_string(mods, key.ptr, submodule->branch);
+	if (error == GIT_ENOTFOUND)
```
Comment (ethomson) [inline src/submodule.c]: Please `giterr_clear` here also.

**Comment by ethomson** [comment]
:+1:  Thanks for this, and sorry about the delay in responding.  Looks good aside from the minor error code clearing issue.

**Comment by ethomson** [comment]
:sparkles:

Thanks!

**Pull request merged**